_applications/bad-apples.adoc

```diff
@@ -0,0 +1,310 @@
+
+[[introduction]]
+== Introduction
+
+The Bad Apples project demonstrates a workflow for creating services that
+combine the distributed processing features of
+https://spark.apache.org[Apache Spark] with the
+https://en.wikipedia.org/wiki/Business_rule[business rules] capabilities of
+http://drools.org/[Drools]. This tutorial will walk you through the process of
+deploying an application that utilizes data analysis and distributed rules
+processing to select a set of possibly fraudulent credit card transactions.
+
+The data used in this tutorial is
+the https://www.kaggle.com/mlg-ulb/creditcardfraud[Credit Card Fraud Detection]
+set from https://kaggle.com[Kaggle].  This set has been chosen to help
+illustrate an example of a real-world problem which requires machine processing
+and human domain expertise. Although the data contains columns which represent
+unknown principal components for the transactions, we have taken these as
+inspiration to demonstrate the types of algorithms and rules you will create
+with your data.
+
+The techniques and algorithms demonstrated in this tutorial are **not**
+recommended for production work where real financial records are involved. They
+are exposed here to illustrate how the underlying infrastructure and
+frameworks can be used to build these type of systems.
+
+[[architecture]]
+== Architecture
+
+There are three primary services which comprise the Bad Apples suite: a
+database, a Spark cluster, and the filter
+https://docs.okd.io/latest/dev_guide/jobs.html[Job]. To aid this tutorial there are
+two additional services you will use: one to load the database, and one to
+observe the results of the filter processing.
+
+The filter job is an application that will extract credit card transaction
+data from the database, filter out transactions above a predefined amount, and
+then process the filtered transactions through the business rules to identify
+suspected fraudulent transactions. The business rules are crafted to select
+transactions based on one of the principal component values.
+
+In practice, the flow of data looks like this:
+
+pass:[<img src="/assets/howdoi/bad-apples.svg" alt="Bad Apples architecture" class="img-responsive">]
+
+Of the two non-primary services, the watcher is shown in this diagram to
+illustrate its place in the architecture. The watcher is a simple service
+that periodically reads from the database to display the number of detected
+fraudlent transactions. Using this service you can easily monitor how a change
+to the filter job affects the number of detected transactions.
+
+=== Record selection process
+
+The filter job service will examine all the records in the database and use
+a few simple algorithms to select the fraudulent transaction records. To better
+understand this process, let's begin by looking at the schema for the
+credit card transaction data.
+
+....
+CREATE TABLE transactions
+(
+Time VARCHAR,
```

Comment:
Would it be easier on the eye to have (say) 3 or 4 column names per row (since the 30 lines are very similar)? e.g.
```
V1 VARCHAR, V2 VARCHAR, V3 VARCHAR, V4 VARCHAR,
V5 VARCHAR, V6 VARCHAR, V7 VARCHAR, V8 VARCHAR,
...
```

Reply:
good suggestion, my main concern is not having the couple of columns we want to highlight get lost in the mix. i'll try reformatting to use multiple columns. thanks!